tests/test_multidict_.py

```diff
@@ -0,0 +1,133 @@
+import pytest
+
+from multidict import MultiDict, CIMultiDict, MultiDictProxy, CIMultiDictProxy
+
+MUTABLE_DICTS = [MultiDict, CIMultiDict]
+PROXY_DICTS = [MultiDictProxy, CIMultiDictProxy]
+
+
+@pytest.fixture(params=MUTABLE_DICTS)
+def md_simple(request):
+    return request.param([('key1', 'value1')])
+
+@pytest.fixture(params=MUTABLE_DICTS)
+def md_w_multivalue_per_key(request):
+    return request.param([('key1', 'one'), ('key2', 'two'), ('key1', 3)])
+
+
+@pytest.mark.parametrize("cls", MUTABLE_DICTS)
+class TestInstantiation:
+
+    def test_instantiate__empty(self, cls):
+        d = cls()
+        assert d == {}
+        assert len(d) == 0
+        assert list(d.keys()) == []
+        assert list(d.values()) == []
+        assert list(d.items()) == []
+
+        assert cls() != list()
+        with pytest.raises(TypeError, match='\(2 given\)'):
+            cls(('key1', 'value1'), ('key2', 'value2'))
+
+    @pytest.mark.parametrize('arg0', [
+        [('key', 'value1')],
+        {'key': 'value1'}
+    ])
+    def test_instantiate__from_arg0(self, cls, arg0):
+        d = cls(arg0)
+
+        assert d == {'key': 'value1'}
+        assert len(d) == 1
+        assert list(d.keys()) == ['key']
+        assert list(d.values()) == ['value1']
+        assert list(d.items()) == [('key', 'value1')]
+
+    def test_instantiate__with_kwargs(self, cls):
+        d = cls([('key', 'value1')], key2='value2')
+
+        assert d == {'key': 'value1', 'key2': 'value2'}
+        assert len(d) == 2
+        assert sorted(d.keys()) == ['key', 'key2']
+        assert sorted(d.values()) == ['value1', 'value2']
+        assert sorted(d.items()) == [('key', 'value1'),
+                                     ('key2', 'value2')]
+
+    def test_instantiate__from_generator(self, cls):
+        d = cls((str(i), i) for i in range(2))
+
+        assert d == {'0': 0, '1': 1}
+        assert len(d) == 2
+        assert sorted(d.keys()) == ['0', '1']
+        assert sorted(d.values()) == [0, 1]
+        assert sorted(d.items()) == [('0', 0), ('1', 1)]
+
+    def test_cannot_create_from_unaccepted(self, cls):
+        with pytest.raises(TypeError):
+            cls([(1, 2, 3)])
+
+
+@pytest.mark.parametrize("cls", MUTABLE_DICTS+PROXY_DICTS)
+class TestContents:
+
+    @pytest.fixture(autouse=True)
+    def _autoassign_md_to_d(self, md_w_multivalue_per_key):
```

Comment:
I like the parametrization approach but this particular fixture looks like a hack.

Instead of `self.d` we can always use a fixture for multidict instance explicitly.
This allows getting rid of `self.d` entirely.
`self.d` is necessary for `unittest` based approach but redundant for pytest

Reply:
Also, classes are normally not necessary. If you need namespacing, flat is better than nested: just use separate test modules :)

Reply:
Agree but I don't insist.
Can live with classes if they are already exist.

Reply:
Yeah, but classes often implicitly encourage ppl to have state. Which might lead to tests depending on each other. Why rely on human reviewers to catch those when simple rules are more effective in terms of discouraging this? :)

Reply:
Is getting rid of classes the suggestion or requirement?

Reply:
Suggestion. A requirement would need actual linting implementation.

Reply:
(But still, it's also: simple is better than complex)

Reply:
@webknjaz regarding classes -> modules part:
I mostly agree but it is hard to split something in modules when you don't actually know all parts.
Like here I am still in progress figuring out the layout and all needed fixtures. 
When it is settled - I will probably come back to that thought.